### PR TITLE
SplitProvider - Functional and aware of config changes

### PR DESCRIPTION
**Pull request by CarsonF**
- Changed `client` to be nullable.
  Now we don't have to fake a default client (`{} as SplitIO.Client`).
  This will prevent error if client is not defined and we try to call `getTreatmentsWithConfig`.
- Changed `lastUpdate` to just be `0` instead of `null` for default.
   Since it is a unix timestamp check 0 functions fine.
- I'm also setting `lastUpdate` to "now" when the client is ready, because I'm pretty sure that's the time it has the treatments loaded from server.
- I'm re-creating the client if the config options have changed.
  By giving all the config options as a dependency list to `useEffect`.
- Converted `SplitProvider` to a function component.

### CODE REVIEW

File: src/SplitProvider.tsx
Comment (samuelcastro) [inline src/SplitProvider.tsx]: I'd create a function to handle this, it's confused to understand what's the purpose of this function, to keep a good readability is a bad idea use generics names like `k` and `v`.

```
const getFeatureEntries = (features: FeaturesType) => Object.entries(features).map((feature, index) => `${feature}::${index}`)
```

Btw, why do we need it?

File: src/SplitProvider.tsx
Comment (samuelcastro) [inline src/SplitProvider.tsx]: What's `clientDeps` and why do we need it?

File: src/SplitProvider.tsx
Comment (samuelcastro) [inline src/SplitProvider.tsx]: The default value of `startup` is an empty object, why? Will there be a case that `startup` will be null and it needs to be set to `{}`? If that's the case, you're trying to get different values out of it below, this will potentially throw an exception when `startup` be just `{}`

**Review by samuelcastro — CHANGES_REQUESTED** [review]
Looks great, just a few small details.

File: src/SplitProvider.tsx
Comment (CarsonF) [reply src/SplitProvider.tsx]: Correct the goal is just to get the values out of it below. It is an optional property so if it's omitted we default it to object here. null isn't valid and TS (if used) would catch that. As long as it is an object either passed in as one or default here, we can pull properties (whose values may be undefined - that's ok) without exceptions.

File: src/SplitProvider.tsx
Comment (CarsonF) [reply src/SplitProvider.tsx]: Yeah I'm not particularly thrilled with that either. I gave it no thought, just was trying to convert the object to a flat list for dependencies (see other comment).

File: src/SplitProvider.tsx
Comment (CarsonF) [reply src/SplitProvider.tsx]: To re-create the client when config changes and to _not_ re-create the client when config _doesn't_ change.

`useEffect` uses this list as a way to memoize the effect called. So if the dependency list is the same as it was the last time the function was ran, it doesn't call the function again.

It does identity checks on each value in the list, which unfortunate since config is an object.
Take this for example:
```tsx
<SplitProvider
  config={{
    core: {
      key: 'my key'
    }
  }}
>
  ...
</SplitProvider>
```
Every time the component including ^ that re-renders `SplitProvider` is passed a _new_ object for its config property.
If our create client logic looked like this
```tsx
useEffect(() => {
  // create client
}, [config]);
```
Then the effect would be called every render as the config object's identity would be different.

So we get around this by checking all the values in the config object as scalars instead of object itself. It sucks, but I don't think we should put it on the consumers to memoize their config objects.

Open to suggestions about different ways to do this too. I wonder if there is a more performant way.

File: src/SplitProvider.tsx
Comment (CarsonF) [inline src/SplitProvider.tsx:16]: I realized I kept both your `isReady` and my client nullable. I don't think both of these are necessary. I've already given my reasons for making client nullable (and still hiding that fact in our consumer components/hooks). And `isReady` can be the same as `client != null`. Thoughts?

File: src/SplitProvider.tsx
Comment (CarsonF) [inline src/SplitProvider.tsx]: I just got a random error about setting state while the component was not mounted. I think we need to check that the component is still mounted before setting state since this is an async call that could happen at random times. I'll look into this.

**Comment by CarsonF** [comment]
Bleh I did a review instead of single comments 🤦‍♂

File: src/SplitProvider.tsx
Comment (samuelcastro) [reply src/SplitProvider.tsx]: We could use `JSON.stringify` to do a deep-equality check, which in terms of performance is similar to what you're doing, just more readable, but also more secure, let's say another prop is injected in the split config object, we would need to update our code to reflect that. Thoughts?

File: src/SplitProvider.tsx
Comment (samuelcastro) [reply src/SplitProvider.tsx:16]: `isReady` is a SDK related attribute not necessarily whether `client` is null or not

File: src/SplitProvider.tsx
Comment (samuelcastro) [reply src/SplitProvider.tsx]: Ok.

File: src/SplitProvider.tsx
Comment (CarsonF) [reply src/SplitProvider.tsx:16]: True....don't think we could hide this though?

Comment (samuelcastro) [reply src/SplitProvider.tsx:16]: We could, wouldn't be good to share when the SDK is ready thou?

Comment (CarsonF) [reply src/SplitProvider.tsx:16]: I'm not sure. The docs say that all split values return `control` as the default value in the case that something fails or is not ready. My hook does the same thing (I'm not sure if the `Split` component got updated with that though).
If the implementor is trying to exhaustively handle all cases it seems that "not ready" is the same as `"control"` value.
https://help.split.io/hc/en-us/articles/360020528072-Control-treatment

If you don't agree I'll back down. It would only matter to people if they are using the context directly instead of our hooks / HOC.

File: src/SplitProvider.tsx
Comment (CarsonF) [reply src/SplitProvider.tsx]: > also more secure, let's say another prop is injected in the split config object, we would need to update our code to reflect that.

Right, exactly why I think this is hacky.

`JSON.stringify` also seems hacky to me too though. What if the key order changes and then the comparison string is different? I would be more comfortable with lodash.isEqual...

Comment (samuelcastro) [reply src/SplitProvider.tsx]: I just don't want to add any other libraries, I think `JSON.stringify` is a great option, I don't think it's hacky, `isEqual` is not even available in recent versions, now it's `isEqualWith`. I'm not quite sure I understood "what if the key order change", from the user perspective or from split.io sdk perspective you mean?

File: src/SplitProvider.tsx
Comment (samuelcastro) [reply src/SplitProvider.tsx:16]: Is this a new feature in the SDK? Because before when the SDK was not ready we used to get a `treatment: null`. But even returning a default `control` treatment, users might not know what that is, the `isReady` would be a simple (maybe dumb) way to know that.

File: src/SplitProvider.tsx
Comment (samuelcastro) [reply src/SplitProvider.tsx]: Have you managed to fix the error?

File: src/SplitProvider.tsx
Comment (CarsonF) [reply src/SplitProvider.tsx]: Yeah I hear you. It adds complexity. Where do you see `isEqual` not available?

`JSON.stringify` will fail like this:
```ts
'{foo: "red", bar: "blue"}' === '{bar: "blue", foo: "red"}'
```
If we just want to note this as a warning and move on I guess that's fine.

That also wouldn't work for the config options that are functions, like `impressionListener`.

File: src/SplitProvider.tsx
Comment (CarsonF) [reply src/SplitProvider.tsx]: I haven't ran into it again. It has to do with HMR because that's the only time that the component is unmounted. I had stubbed out a potential fix that used [`react-use's useMountedState`](https://github.com/streamich/react-use/blob/master/docs/useMountedState.md) to check if the component was mounted before calling the setters.

File: src/SplitProvider.tsx
Comment (samuelcastro) [reply src/SplitProvider.tsx]: I was reading the lodash source code (https://github.com/lodash/lodash), I couldn't find it there, I think I'm wrong the documentation still has it.

I understand that order will change the results of `JSON.stringify` I guess my question is, in which context the order of split.io config object might change, once it's setup I doubt users will keep changing its order. The function is irrelevant I think, we're just trying to do a deep equality check.

But I think either your solution or `JSON.stringify` is fine, I'll let you decide.

File: src/SplitProvider.tsx
Comment (CarsonF) [reply src/SplitProvider.tsx:16]: I'm not sure if it is a new feature. I just saw it in the docs. I'm not sure it is this libraries job to inform users how Split works. We are bound by their constraints after all.

I'm fine keeping it. At this point I'd rather give implementors more information than less. This `isReady` can easily be hidden by a hook or HOC. And it even is in my hooks, but we can talk about that in that PR not here.

File: src/SplitProvider.tsx
Comment (CarsonF) [reply src/SplitProvider.tsx]: I think lodash is a better solution but I'm hesitant to go with it because of what it could imply with compilation. I'm really weak with what's needed to compile and publish TS/JS libraries.

We can always optimize down the road.

Comment (samuelcastro) [reply src/SplitProvider.tsx]: I agree, are you going to leave it as it is or use `JSON.stringify`?

File: src/SplitProvider.tsx
Comment (samuelcastro) [reply src/SplitProvider.tsx:16]: Sounds good to me.

File: src/SplitProvider.tsx
Comment (CarsonF) [reply src/SplitProvider.tsx]: I'll change it to `JSON.stringify`. Maybe I can wrap it in a memo call so we don't incur the `JSON.stringify` cost if the object identity hasn't changed.

**Comment by CarsonF** [comment]
Ok I'm ready for another review @samuelcastro. I did everything we've talked about so far.

I also added custom handling for ~`debug` and~ `impressionListener` settings. You can see why in code/commits. If you don't like these or don't think it is necessary, then I can remove.

EDIT: I removed the `debug` special handling because it was destructuring the config property which would _always_ give it a new identity.

**Comment by CarsonF** [comment]
I thought of another caveat. Since we are memoize the config object before `JSON.stringify` if the user tries to mutate the object the changes won't be applied. Mutating the same object isn't idiomatic for react, but it would be nice not to "do nothing". Maybe `Object.freeze` can be used...

**Comment by samuelcastro** [comment]
The changes looks great, I like the idea of freezing the object, this way users will know mutating is a bad idea. Or we could use [immer](https://github.com/immerjs/immer) and this mutation would never happen, however once again, I'm avoid unnecessary dependencies.

**Review by samuelcastro — APPROVED** [review]
🥇

**Comment by samuelcastro** [comment]
I'm going to release it as an alpha version after merge and then after make sure everything is good to go we can make a final release. If you want to test it as well I have a Code Sandbox live where we can easily test it: https://codesandbox.io/s/jovial-banzai-3bc81

**Comment by CarsonF** [comment]
Ok I added the freeze call. I'm now thinking though that we need to tweak again.
```tsx
<SplitProvider config={ core: { authorizationKey: '' } } />
```
This will run the recursive deep freeze and `JSON.stringify` on every render.
I'm not sure that's a good thing.

The only solution might be to tell users to memoize their config or define outside of component.

**Comment by samuelcastro** [comment]
Yeah I don't like idea either, maybe we should leave it as it is and just let users know mutation will produce unexpected results with the client not being re-created.

**Comment by CarsonF** [comment]
I liked it more when I didn't think I had to implement recursion myself.

I still think we should tell users the best way to write it so we don't have to do `JSON.stringify` on every render. And if we are doing that, and assuming users follow it because they care about performance, then the freeze is also not called needlessly. In that case it would be safer to keep the freeze.

File: README.md
Comment (samuelcastro) [inline README.md]: `React.memo` expects a react component not a function, I think you meant `useMemo` which would make sense here.

**Review by samuelcastro — CHANGES_REQUESTED** [review]

File: README.md
Comment (CarsonF) [reply README.md]: 🤦‍♂

**Review by samuelcastro — APPROVED** [review]
Looks good.

**Comment by CarsonF** [comment]
Ship it! :shipit: 😁 

I'll rebase the other PR next and update docs.

**Pull request merged**
